File: all/data structure/Reservoir Sampling/ReservoirSampling.java
package temp;

import java.util.Random;

/**
 * reference：http://www.cnblogs.com/HappyAngel/archive/2011/02/07/1949762.html
 * @author datawlb
 *
Init : a reservoir with the size： k
for i= k+1 to N
    M=random(1, i);
    if( M < k)
     SWAP the Mth value and ith value
end for 
 */
public class ReservoirSampling {

	public static int[] reservoirSample(int[] data, int k){
		if((data == null)||data.length <= k)
			return new int[0];
		int[] sampleResult = new int[k];
		for(int i=0; i<k; i++)
			sampleResult[i] = data[i];
		for(int i=k; i<data.length; i++){
			int random = new Random().nextInt(i+1);
			if(random < k)
				sampleResult[random] = data[i];
		}
		return sampleResult;
	}
	public static void main(String[] args){
		int k = 50;
		int n = 1000;
		int[] data = new int[n];
		for(int i=0; i<n; i++)
			data[i] = i;
		int[] result = reservoirSample(data, k);
		System.out.println(result.length);
	}
}

File: all/data structure/Reservoir Sampling/ReservoirSampling.scala
package org.test

import scala.util.Random

/**
 * Created by datawlb on 2015/5/10.
 */
object ReservoirSampling {
  def reservoirSample(data: Array[Int], k: Int): Array[Int] = {
    val result = new Array[Int](k)
    for(i<-0 until(data.length)){
      if (i<k)
        result(i) = data(i)
      else{
        val rand1 = new Random().nextInt(i+1)
        if(rand1 < k)
          result(rand1) = data(i)
      }
    }
    result
  }
  def main (args: Array[String]) {
    val k = 50
    val n = 1000
    val data = new Array[Int](n)
    for(i<-0 until(n))
      data(i) = i
    println(reservoirSample(data, k).length)
  }
}
